text/0000-read-buf.md

```diff
@@ -0,0 +1,655 @@
+                buf.reserve(32);
+            }
+
+            let mut read_buf = ReadBuf::uninit(buf.spare_capacity_mut());
+            unsafe {
+                read_buf.assume_init(initialized);
+            }
+
+            match self.read_buf(&mut read_buf) {
+                Ok(()) => {}
+                Err(e) if e.kind() = io::ErrorKind::Interrupted => continue,
+                Err(e) => return Err(e),
+            }
+
+            if read_buf.filled().is_empty() {
+                break;
+            }
+
+            initialized = read_buf.initialized().len() - read_buf.filled().len();
+            let new_len = buf.len() + read_buf.filled().len();
+            unsafe {
+                buf.set_len(new_len);
+            }
+        }
+
+        Ok(buf.len() - initial_len)
+    }
+}
+
+pub fn copy<R, W>(reader: &mut R, writer: &mut W) -> io::Result<u64>
+where
+    R: Read,
+    W: Write,
+{
+    let mut buf = [MaybeUninit::uninit(); 4096];
+    let mut buf = ReadBuf::uninit(&mut buf);
+    let mut len = 0;
+
+    loop {
+        match reader.read_buf(&mut buf) {
+            Ok(()) => {},
+            Err(e) if e.kind() == io::ErrorKind::Interrupted => continue,
+            Err(e) => return Err(e),
+        };
+
+        if buf.filled().is_empty() {
+            break;
+        }
+
+        len += buf.filled().len() as u64;
+        writer.write_all(buf.filled())?;
+        buf.clear();
+    }
+
+    Ok(len)
+}
+```
+
+The existing `std::io::Initializer` type and `Read::initializer` method will be removed.
+
+Vectored reads use a similar API:
+
+```rust
+/// A possibly-uninitialized version of `IoSliceMut`.
+///
+/// It is guaranteed to have exactly the same layout and ABI as `IoSliceMut`.
+pub struct MaybeUninitIoSliceMut<'a> { ... }
+
+impl<'a> MaybeUninitIoSliceMut<'a> {
+    /// Creates a new `MaybeUninitIoSliceMut` from a slice of maybe-uninitialized bytes.
+    #[inline]
+    pub fn new(buf: &'a mut [MaybeUninit<u8>]) -> MaybeUninitIoSliceMut<'a> { ... }
+}
+
+impl<'a> Deref for MaybeUninitIoSliceMut<'a> {
+    type Target = [MaybeUninit<u8>];
+
+    ...
+}
+
+impl<'a> DerefMut for MaybeUninitIoSliceMut<'a> { ... }
+
+
+/// A wrapper over a set of incrementally-initialized buffers.
+pub struct ReadBufs<'a> { ... }
+
+impl<'a> ReadBufs<'a> {
+    /// Creates a new `ReadBufs` from a set of fully initialized buffers.
+    #[inline]
+    pub fn new(bufs: &'a mut [IoSliceMut<'a>]) -> ReadBufs<'a> { ... }
+
+    /// Creates a new `ReadBufs` from a set of fully uninitialized buffers.
+    ///
+    /// Use `assume_init` if part of the buffers are known to be already initialized.
+    #[inline]
+    pub fn uninit(bufs: &'a mut [MaybeUninitIoSliceMut<'a>]) -> ReadBufs<'a> { ... }
+
+    ...
+}
+
+pub trait Read {
+    /// Pull some bytes from this source into the specified set of buffers.
+    ///
+    /// This is equivalent to the `read_vectored` method, except that it is passed a `ReadBufs` rather than
```

Comment:
```suggestion
    /// This is equivalent to the `read_vectored` method, except that it is passed a `ReadBuf` rather than
```

Reply:
I don’t think this is a typo; there are two similarly named types. The plural version appears to be used for vectored IO.

Reply:
Yep, this is not a typo.

Reply:
Why do we need to distinct this, doesn't this confuse others? Well, I got confused without reading the whole thing.

Reply:
The whole point of vectored IO is to be able to work with multiple non-contiguous buffers in a single call. `ReadBuf` works with a single buffer, so it can't be used in a vectored method.

Reply:
Can it have a better name that is less confused when used together with `ReadBuf`?

Reply:
I suggest `ReadBufVectored` or `ReadBufVec` instead of `ReadBufs`